mpes/fprocessing.py

```diff
@@ -2297,6 +2455,329 @@ def distributedBinning(self, axes, nbins, ranges, binDict=None, pbar=True,
+                break
+        else:
+            print("Lens mode for given lens voltages not found.\n \
+            Storing lens mode from the user, if provided.")
+            
+        # Determining projection from the lens mode
+        try:
+            lens_mode = metadata_dict["instrument"]["analyzer"]["lens_mode"]
+            if "spatial" in lens_mode.split("_")[1]:
+                metadata_dict["instrument"]["analyzer"]["projection"] = "real"
+            else:
+                metadata_dict["instrument"]["analyzer"]["projection"] = "reciprocal"
+        except IndexError:
+            print("Lens mode must have the form, '6kV_kmodem4.0_20VTOF_v3.sav'.\
+                Can't determine projection.\n \
+                Storing projection from the user, if provided.")
+        except KeyError:
+            print("Lens mode not found. Can't determine projection.\n \
+                Storing projection from the user, if provided.")
+
+        default_units = {
+        'X': 'step', 
+        'Y': 'step', 
+        't': 'step', 
+        'tofVoltage':'V',
+        'extractorVoltage':'V',
+        'extractorCurrent':'A',
+        'cryoTemperature':'K',
+        'sampleTemperature':'K',
+        'dldTimeBinSize':'ns',
+        'delay':'ps',
+        'timeStamp':'s',
+        'E':'eV',
+        'kx':'1/A',
+        'ky':'1/A'}
+
+        def res_to_xarray(res, binNames, binAxes, metadata=None):
+            """ creates a BinnedArray (xarray subclass) out of the given np.array
+            Parameters:
+                res: np.array
+                    nd array of binned data
+                binNames (list): list of names of the binned axes
+                binAxes (list): list of np.arrays with the values of the axes
+            Returns:
+                ba: BinnedArray (xarray)
+                    an xarray-like container with binned data, axis, and all available metadata
+            """
+            dims = binNames
+            coords = {}
+            for name, vals in zip(binNames, binAxes):
+                coords[name] = vals
+
+            xres = xr.DataArray(res, dims=dims, coords=coords)
+
+            for name in binNames:
+                try:
+                    xres[name].attrs['unit'] = default_units[name]
+                except KeyError:
+                    pass
+
+            xres.attrs['units'] = 'counts'
+            xres.attrs['long_name'] = 'photoelectron counts'
+
+            if metadata is not None:
+                xres.attrs['metadata'] = metadata
+
+            return xres
+
+        print("Building the xarray data object...")
+        axnames = self.binaxes.copy()
+        axnames[2] = "energy"
+        axes = [self.histdict[ax] for ax in self.binaxes]
+        res_xarray = res_to_xarray(self.histdict['binned'], axnames, axes, metadata=metadata_dict)
+        print("Done!")
+        return res_xarray
+
+    def xarray_to_h5(self, data, faddr, mode='w'):

```

Comment:
Move the function out of the class, and add a class wrapper around it, that accesses the class attributes.

Reply:
What I mean is to move the function out of the class, and then add a function to the class for saving e.g. self.xarray ore something like this, without needing to supply the data explicitly.